### PR TITLE
Improved Delta concurrency with finer-grained conflict detection in OptTxnImpl

### PR DESCRIPTION
**Pull request by tdas**
This is a modified PR from the original PR https://github.com/delta-io/delta/pull/114 by @tomasbartalos (kudos, it was a very good PR!). This PR tracks transaction changes at a finer granularity (no new columns required in RemoveFile action) thus allowing more concurrent operations to succeed.

closes delta-io/delta#228 and delta-io/delta#72

## What changes were proposed in this pull request?

This PR improves the conflict detection logic in OptTxn using the following strategy. 
- OptTxn tracks two additional things 
  - All the partitions read by the query using the OptTxn
  - All the files read by the query
- When committing a txn, it checks this txn's actions against the actions of concurrently committed txns using the following strategy:
  1. If any of the concurrently added files are in the partitions read by this txn, then fail because this txn should have read them. 
      -It’s okay for files to have been removed from the partitions read by this txn as long as this txn never read those files. This is checked by the next rule.
  2. If any of the files read by this txn have already been removed by concurrent txns, then fail.
  3. If any of the files removed by this txn have already been removed by concurrent txns, then fail.
- In addition, I have made another change where setting `dataChange` to `false` in all the actions (enabled by #223) will ensure the txn will not conflict with any other concurrent txn based on predicates.

## How was this patch tested?
Tests written by @tomasbartalos in the original PR. Some tests were changed because some scenarios that were blocked in the original PR are now allowed, thanks to more granular and permissive conflict detection logic. Some test names tweaked to ensure clarity.

Co-authored-by: Tathagata Das <tathagata.das1565@gmail.com>
Co-authored-by: Tomas Bartalos <tomas.bartalos@nike.sk>

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 1]
```diff
@@ -145,6 +145,12 @@ trait OptimisticTransactionImpl extends TransactionalWrite {
   /** Tracks the appIds that have been seen by this transaction. */
   protected val readTxn = new ArrayBuffer[String]
 
+  /** Tracks the data that could have been seen by this transaction. */
+  protected val readPredicates = new ArrayBuffer[Expression]
```
Comment (mukulmurthy) [inline src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala]: I'd rephrase this to "Used to track the data ..."

File: src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 2]
```diff
@@ -446,22 +483,89 @@ trait OptimisticTransactionImpl extends TransactionalWrite {
...
+      // Fail if new files have been added that the txn should have read.
+      val addedFilesToCheckForConflicts = commitIsolationLevel match {
+        case Serializable => changedDataAddedFiles ++ blindAppendAddedFiles
+        case WriteSerializable => changedDataAddedFiles // don't conflict with blind appends
+        case SnapshotIsolation => Seq.empty
       }
+      val predicatesMatchingAddedFiles = ExpressionSet(readPredicates).iterator.flatMap { p =>
+        val conflictingFile = DeltaLog.filterFileList(
+          metadata.partitionColumns,
+          addedFilesToCheckForConflicts.toDF(), p :: Nil).as[AddFile].take(1)
```
Comment (mukulmurthy) [inline src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala]: nit: why take(1) instead of just headOption here?

File: src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 3]
```diff
@@ -446,22 +483,89 @@ trait OptimisticTransactionImpl extends TransactionalWrite {
...
+
+        conflictingFile.headOption.map(f => getPrettyPartitionMessage(f.partitionValues))
+      }.take(1).toArray
+
+      if (predicatesMatchingAddedFiles.nonEmpty) {
+        val isWriteSerializable = commitIsolationLevel match {
+          case WriteSerializable => true
+          case _ => false
+        }
```
Comment (mukulmurthy) [inline src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala]: nit: should we just do commitIsolationLevel == WriteSerializable below? this is pretty verbose

File: src/main/scala/org/apache/spark/sql/delta/isolationLevels.scala [hunk 4]
```diff
@@ -0,0 +1,91 @@
+/*
```
Comment (mukulmurthy) [inline src/main/scala/org/apache/spark/sql/delta/isolationLevels.scala]: should this file be called IsolationLevel.scala?

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala [hunk 5]
```diff
@@ -93,6 +96,581 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
...
+      tx1.commit(addA_P1 :: Nil, Truncate())
+
+      intercept[ConcurrentModificationException] {
+        // tx1 modified P1 which we tried to read and could try to delete
+        tx2.commit(Seq.empty, Truncate())
+      }
+    }
+  }
+  */
+
```
Comment (mukulmurthy) [inline src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: what's this?

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala [hunk 6]
```diff
@@ -93,6 +96,581 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
...
+      test: DeltaLog => Unit): Unit = {
+
+    val schema = new StructType(partitionCols.map(p => new StructField(p, StringType)).toArray)
+    var actionWithMetaData =
+      actions :+ Metadata(partitionColumns = partitionCols, schemaString = schema.json)
+
+    withTempDir { tempDir =>
+      val log = DeltaLog(spark, new Path(tempDir.getCanonicalPath))
+      // Initialize the log and add data. Truncate() is just a no-op placeholder.
+      log.startTransaction().commit(actionWithMetaData, Truncate())
```
Comment (mukulmurthy) [inline src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: Why Truncate()?

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala [hunk 7]
```diff
@@ -93,6 +96,581 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
...
+      val tx1Read = tx1.filterFiles(('part === 1).expr :: Nil)
+      assert(tx1Read.isEmpty)
+
+      // tx2 deletes "part=p1"
+      val tx2 = log.startTransaction()
+      // empty read
+      val tx2Read = tx2.filterFiles(('part === 1).expr :: Nil)
+      assert(tx2Read.isEmpty)
+
+      tx1.commit(addD_P2 :: Nil, Truncate())
```
Comment (zsxwing) [inline src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: Should be `addD_P2.remove` since `truncate` should just remove files.

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala [hunk 8]
```diff
@@ -93,6 +96,581 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
...
+    }
+  }
+  */
+
+  test("allow concurrent commit on Delete and Add disjoint partitions initially empty") {
+    withLog(addC_P2 :: addE_P3 :: Nil) { log =>
+      // tx1 deletes "part=p1"
+      val tx1 = log.startTransaction()
+      // empty read
+      val tx1Read = tx1.filterFiles(('part === 1).expr :: Nil)
```
Comment (zsxwing) [inline src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: I don't get what's this test for. I would expect if there was a truncate table operation, it would always touch the whole table.

If this test is for `replaceWhere`, the right operation here should be something like `DeltaOperations.Write(mode, Option(partitionColumns), options.replaceWhere)`.

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala (hunk 7 above)
Comment (tdas) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: looking at the existing tests, i think Truncate() is just a place holder operation. could have been anything. 

we should probably change all of it to use ManualUpdate

File: src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala (hunk 2 above)
Comment (tdas) [reply src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala]: no reason. no difference.

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala (hunk 5 above)
Comment (tdas) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: was supposed to be removed. incorrect test.

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala (hunk 8 above)
Comment (tdas) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: I think I missed fixing this test.

Comment (tdas) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: yeah .i dont think this test really tests anything that has not been covered by other tests. I am removing this.

File: src/main/scala/org/apache/spark/sql/delta/isolationLevels.scala (hunk 4 above)
Comment (tdas) [reply src/main/scala/org/apache/spark/sql/delta/isolationLevels.scala]: the recommended style from the Spark world is that if there are multiple related top-level classes (that is, public classes used outside the file) are present in the same file, then it should be named not by any of the class but as common name across them starting with small letter. Example - https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala

Comment (mukulmurthy) [reply src/main/scala/org/apache/spark/sql/delta/isolationLevels.scala]: Ha, me and Andreas were just talking about this the other day. I thought that if there's one main one or it's an abstract class/trait and classes that extend it, it's named after the main one. But I guess this one could fit into either case. Either way fine with me

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala [hunk 9]
```diff
@@ -93,6 +96,581 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
...
+      // TX1 reads only P1
+      val tx1Read = tx1.filterFiles(('part === 1).expr :: Nil)
+      assert(tx1Read.map(_.path) == A_P1 :: Nil)
+
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      // TX2 modifies only P1
+      tx2.commit(addB_P1 :: Nil, Truncate())
+
+      intercept[ConcurrentModificationException] {
```
[the remark this answers is not in the document]
Comment (tdas) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: done

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala [hunk 10]
```diff
@@ -34,76 +37,632 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
...
+
+  test("allow concurrent commit on disjoint partitions reading all partitions") {
+    withLog(addA_P1 :: addD_P2 :: addD_P2.remove :: Nil) { log =>
+      val tx1 = log.startTransaction()
+      // TX1 read P1
+      tx1.filterFiles(('part isin 1).expr :: Nil)
+
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      tx2.commit(addC_P2 :: addD_P2.remove :: Nil, ManualUpdate)
```
Comment (cfmcgrady) [inline src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: this file `addD_P2` already removed by initial commit, can we remove the same file multi-time?

Comment (tdas) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: That's definitely a mistake. `addD_P2.remove` should not be in the initial commit of the log. In terms of guarantees, technically it is okay for a committed log to have duplicate removes. However, it is not okay for concurrent txns to concurrently attempt to remove the same file multiple times, because it breaks the semantic integrity of the second operation.

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala (hunk 8 above)
Comment (tomasbartalos) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: The intension of the 2 tests ("block concurrent commit on Delete attempt and Add initially empty" and "allow concurrent commit on Delete and Add disjoint partitions initially empty") was to check "empty deletes" and subsequent adds. Empty deletes are deletes whose condition doesn't satisfy any data at the time of execution, however concurrent transactions may commit data that would conflict with the condition. This is a superset of replaceWhere scenario which is tested in "block concurrent replaceWhere initial empty". The difference is that replaceWhere AddFile partitions have to match the delete condition ('part === 1), while in this test the AddFile's partition (part -> 2) is out of delete condition ('part === 1). As mentioned elsewhere Truncate() doesn't have any meaning, it is just a "no op placeholder" and it was correctly changed to ManualUpdate.

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala (hunk 7 above)
Comment (tomasbartalos) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: Correct, the use of Truncate() was just took over from existing tests.

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala (hunk 10 above)
Comment (tomasbartalos) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: True, please also change TX1's read to all partitions: `tx1.filterFiles(('part isin (1, 2)).expr :: Nil)`

Comment (tdas) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: I am not sure I get why? The point of this test is to test that tx1 will be allowed because it is reading only part1 and tx2 is making changes to partition 2 only. If we change tx1 to read both partitions 1 and 2, then this tx1 is supposed to fail .. isnt it?

Comment (tomasbartalos) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: You're right, however this is already checked in "allow concurrent commit on disjoint partitions". The idea behind this one was that TX1 reads all partitions the table has (almost like full table scan), and it is still allowed to commit. To make this correct, the TX2 should commit addG_P4 (not *_P2), so TX1 and TX2 would operate on disjoint partitions.

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala [hunk 11]
```diff
@@ -16,13 +16,16 @@
 
 package org.apache.spark.sql.delta
 
-import org.apache.spark.sql.delta.DeltaOperations.Truncate
-import org.apache.spark.sql.delta.actions.{AddFile, Metadata, SetTransaction}
+import java.util.ConcurrentModificationException
```
Comment (tomasbartalos) [inline src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: nit: optimise imports

File: src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala (hunk 10 above)
Comment (tdas) [reply src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala]: Aah I get it now. That makes sense! Unfortunately, I merged this PR in our internal repo already to speed up things towards the 0.5.0 release. I will create a new internal PR and address this. 

Thank you very much for reviewing and working with us on this :)

**Comment by zsxwing** [comment]
Resolved by f32830022e6a664f99f25f3dad8584f5cd9952bf
